### PR TITLE
Support /sys/devices/system/cpu/online

### LINKED ISSUE #88: Support for /sys/devices/system/cpu
Support for /sys/devices/system/cpu so that the LXC system does not see the additional CPUs available for host.

Current behavior causes bugs with CPU count autodetection on programs like nginx and others which try to generate their process count based on the CPUs available on system. They will use host's value which is wrong.

Example of current, non working, /sys/devices/system/cpu:

```
root@fr-gcp4:/sys/devices/system/cpu# ls
cpu0   cpu12  cpu16  cpu2   cpu23  cpu27  cpu30  cpu6  cpuidle       microcode  possible
cpu1   cpu13  cpu17  cpu20  cpu24  cpu28  cpu31  cpu7  intel_pstate  modalias   power
cpu10  cpu14  cpu18  cpu21  cpu25  cpu29  cpu4   cpu8  isolated      offline    present
cpu11  cpu15  cpu19  cpu22  cpu26  cpu3   cpu5   cpu9  kernel_max    online     uevent
```

Example of how it should work:

```
root@fr-gcp4:/sys/devices/system/cpu# ls
cpu0   cpu1  cpu2  cpu3   cpu4  cpu5 cpuidle       microcode  possible ntel_pstate  modalias 
power isolated      offline    present kernel_max    online     uevent
```

### PR DESCRIPTION
**Pull request by wavezhang**
Support /sys/devices/system/cpu/online， to make applications that use sysconf (_SC_NPROCESSORS_ONLN) happy.

Partially resolve #88 .

### CODE REVIEW

File: bindings.c [hunk 1]
```diff
@@ -5874,6 +5878,275 @@ int proc_read(const char *path, char *buf, size_t size, off_t offset,
...
 
+static int sys_devices_system_cpu_read(char *buf, size_t size, off_t offset,
+		struct fuse_file_info *fi)
+{
+	struct fuse_context *fc = fuse_get_context();
+	struct file_info *d = (struct file_info *)fi->fh;
+	char *cache = d->buf;
+	ssize_t total_len = 0;
+	if (offset){
+		if (!d->cached)
```
Comment (brauner) [inline bindings.c]: - please add a new line after ssiz_t and if (offset)
- please leave a spacer after the closing ) and the opening {

File: bindings.c [hunk 2]
```diff
@@ -5874,6 +5878,275 @@ int proc_read(const char *path, char *buf, size_t size, off_t offset,
...
+	total_len = snprintf(d->buf, d->buflen, "0-1\n");
+	if (total_len < 0 || total_len >=  d->buflen){
+		lxcfs_error("%s\n", "failed to write to cache");
+		return 0;
+	}
+
+	d->size = (int)total_len;
+	d->cached = 1;
+
+	if (total_len > size) total_len = size;
```
Comment (brauner) [inline bindings.c]: please move the statement belonging onto a new line

File: bindings.c [hunk 3]
```diff
@@ -5874,6 +5878,275 @@ int proc_read(const char *path, char *buf, size_t size, off_t offset,
...
+	struct fuse_context *fc = fuse_get_context();
+	struct file_info *d = (struct file_info *)fi->fh;
+	char *cache = d->buf;
+	char *cg;
+	char *cpuset = NULL;
+	bool use_view;
+
+        int max_cpus = 0;
+	ssize_t total_len = 0;
+	if (offset){
```
Comment (brauner) [inline bindings.c]: see comment above

File: bindings.c [hunk 4]
```diff
@@ -5874,6 +5878,275 @@ int proc_read(const char *path, char *buf, size_t size, off_t offset,
...
+		if (!d->cached)
+			return 0;
+		if (offset > d->size)
+			return -EINVAL;
+		int left = d->size - offset;
+		total_len = left > size ? size: left;
+		memcpy(buf, cache + offset, total_len);
+		return total_len;
+	}
+	pid_t initpid = lookup_initpid_in_store(fc->pid);
```
Comment (brauner) [inline bindings.c]: please only declare new variables at the beginning of a new scope not in the middle

File: bindings.c [hunk 5]
```diff
@@ -5874,6 +5878,275 @@ int proc_read(const char *path, char *buf, size_t size, off_t offset,
...
+		total_len = snprintf(d->buf, d->buflen, "0\n");
+	if (total_len < 0 || total_len >=  d->buflen){
+		lxcfs_error("%s\n", "failed to write to cache");
+		return 0;
+	}
+
+	d->size = (int)total_len;
+	d->cached = 1;
+
+	if (total_len > size) total_len = size;
```
Comment (brauner) [inline bindings.c]: see comment further up

File: bindings.c [hunk 6]
```diff
@@ -5874,6 +5878,275 @@ int proc_read(const char *path, char *buf, size_t size, off_t offset,
...
+}
+
+static off_t get_sysfile_size(const char *which)
+{
+	FILE *f = fopen(which, "r");
+	char *line = NULL;
+	size_t len = 0;
+	ssize_t sz, answer = 0;
+	if (!f)
+		return 0;
```
Comment (brauner) [inline bindings.c]: please do not call a function right when you declare the variable if you check the error condition later, instead do
FILE *f;
.
.
.

f = fopen(which, "r");
if (!f)

**Comment by wavezhang** [comment]
@brauner I've update code based on your suggestion.

**Comment by brauner** [comment]
jenkins: test this please

File: bindings.c [hunk 7]
```diff
@@ -5874,6 +5878,283 @@ int proc_read(const char *path, char *buf, size_t size, off_t offset,
...
+}
+
+int sys_read(const char *path, char *buf, size_t size, off_t offset,
+		struct fuse_file_info *fi)
+{
+	struct file_info *f = (struct file_info *) fi->fh;
+
+	switch (f->type) {
+	case LXC_TYPE_SYS_DEVICES:
+		return sys_devices_read(buf, size, offset, fi);
```
Comment (brauner) [inline bindings.c]: This function is not defined anywhere.

File: bindings.c [hunk 8]
```diff
@@ -5874,6 +5878,283 @@ int proc_read(const char *path, char *buf, size_t size, off_t offset,
...
+	case LXC_TYPE_SYS_DEVICES_SYSTEM:
+		return sys_devices_system_read(buf, size, offset, fi);
```
Comment (brauner) [inline bindings.c]: This function is not defined anywhere.

**Comment by wavezhang** [comment]
I've update codes, dirs do not need to implement 'read' method.

**Comment by brauner** [comment]
`clang` is not happy:
```c
sysfs_fuse.c:37:12: warning: unused function 'sys_devices_system_cpu_read'
      [-Wunused-function]
static int sys_devices_system_cpu_read(char *buf, size_t size, off_t offset,
           ^
1 warning generated.
```
and `gcc` agrees:
```c
sysfs_fuse.c:37:12: warning: ‘sys_devices_system_cpu_read’ defined but not used [-Wunused-function]
 static int sys_devices_system_cpu_read(char *buf, size_t size, off_t offset,
            ^
mv -f .deps/liblxcfs_la-cpuset.Tpo .deps/liblxcfs_la-cpuset.Plo
```

**Comment by wavezhang** [comment]
The function `sys_devices_system_cpu_read ` is never used, I've delete it.

**Comment by brauner** [comment]
I've tested it and it seems to work. To have this useable the script
https://github.com/lxc/lxcfs/blob/master/share/lxc.mount.hook.in
needs to gain something like
```shell
if [ -d /var/lib/lxcfs/sys/devices/system/cpu ]; then
    for entry in /var/lib/lxcfs/sys/devices/system/cpu/*; do
        [ -e "${LXC_ROOTFS_MOUNT}/sys/devices/system/cpu/$(basename $entry)" ] || continue
        mount -n --bind $entry ${LXC_ROOTFS_MOUNT}/sys/devices/system/cpu/$(basename $entry)
    done
fi
```

But, note that you're currently only filtering the `online` file while the `cpu` directory will also container `cpu<idx>` directories for each cpu:
```bash
brauner@wittgenstein|~
> ls -al /sys/devices/system/cpu/| grep -iE cpu[[:digit:]]
drwxr-xr-x 10 root root    0 Apr 15 23:34 cpu0
drwxr-xr-x 10 root root    0 Apr 15 23:34 cpu1
drwxr-xr-x 10 root root    0 Apr 15 23:34 cpu2
drwxr-xr-x 10 root root    0 Apr 15 23:34 cpu3
drwxr-xr-x 10 root root    0 Apr 15 23:34 cpu4
drwxr-xr-x 10 root root    0 Apr 15 23:34 cpu5
drwxr-xr-x 10 root root    0 Apr 15 23:34 cpu6
drwxr-xr-x 10 root root    0 Apr 15 23:34 cpu7
```
To virtualize this properly they should be filtered *but* I'm not opposed to punt on this for now if the tools out there currently only care about what the contents of `online` are. But I'm not sure.

@stgraber, @hallyn, thoughts?

**Comment by hallyn** [comment]
@brauner agreed, progress is progress :)

**Comment by wavezhang** [comment]
So, will this PR merged or not ?

**Comment by brauner** [comment]
> So, will this PR merged or not ?

Had quite a few busy weeks. Sorry.

**Comment by brauner** [comment]
Please note that with your PR there's still logic missing in https://github.com/lxc/lxcfs/blob/master/share/lxc.mount.hook.in
as I pointed out above. This means your new features won't be enabled currently.

**Pull request merged**